### PR TITLE
Image collections

### PR DESCRIPTION
**Pull request by jamesgorrie**
Adds collections to the Collections Dynamo DB.
The store is a `ImageId => List[Collection]` store.

I thought we could do a store of row = `ImageId => CollectionId => Collection`, but then when returning the list we would have to do a look up every time, so I thought keeping the list of collections together might be best.

The side-effect of this is that we can have multiple collections in a list (two people having added it), which might actually have a nice side-effect of being able to create a bit of an audit trail.

`POST collection.com/images/:imageId -d '{ "data":["g2", "film"] }'`
```JSON
{
	"data": [{
		"pathId": "g2/art",
		"path": [
			"g2",
			"art"
		],
		"actionData": {
			"author": "test",
			"date": "2015-11-30T21:33:21.956+00:00"
		}
	}, {
		"pathId": "g2/film",
		"path": [
			"g2",
			"film"
		],
		"actionData": {
			"author": "test",
			"date": "2015-12-07T11:23:06.985+00:00"
		}
	}]
}
```

### CODE REVIEW

File: collections/app/controllers/ImageCollectionsController.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  import CollectionsManager.onlyLatest
+
+  val Authenticated = ControllerHelper.Authenticated
+  val dynamo = new DynamoDB(Config.awsCredentials, Config.dynamoRegion, Config.imageCollectionsTable)
+
+  def getCollections(id: String) = Authenticated.async { req =>
+    dynamo.listGet[Collection](id, "collections").map { dynamoEntry =>
+      respond(onlyLatest(dynamoEntry))
+    } recover {
+      case NoItemFound => respondNotFound("No usage rights overrides found")
```
Comment (akash1810) [inline collections/app/controllers/ImageCollectionsController.scala]: `No usage rights overrides found` copy + paste?

Comment (jamesgorrie) [inline collections/app/controllers/ImageCollectionsController.scala]: :eyes: changed

File: collections/app/controllers/ImageCollectionsController.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      dynamo.listAdd(id, "collections", collection)
+        .map(publish(id))
+        .map(cols => respond(collection))
+    } getOrElse Future.successful(respondError(BadRequest, "invalid-form-data", "Invalid form data"))
+  }
+
+  def publish(id: String)(collections: List[Collection]): List[Collection] = {
+    val message = Json.obj(
+      "id" -> id,
+      "data" -> Json.toJson(onlyLatest(collections))
```
Comment (theefer) [inline collections/app/controllers/ImageCollectionsController.scala]: Do we want to send a diff (like we do for exports) or a full snapshot (like we do for usage and metadata edits)? I'd err towards the latter, else we need another special message for deletions and we are more at risk of concurrency issues (e.g. out of order messages).

File: common-lib/src/test/scala/com/gu/mediaservice/lib/aws/DynamoDBTest.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package com.gu.mediaservice.lib.aws
+
+import com.amazonaws.services.dynamodbv2.document.utils.ValueMap
+import com.gu.mediaservice.model.{ActionData, Collection}
+import org.joda.time.DateTime
+import org.scalatest.{FunSpec, Matchers}
+import play.api.libs.json.{Format, JsObject, Json}
+
+class DynamoDBTest extends FunSpec with Matchers {
```
Comment (theefer) [inline common-lib/src/test/scala/com/gu/mediaservice/lib/aws/DynamoDBTest.scala]: Nice to have!

Comment (jamesgorrie) [inline common-lib/src/test/scala/com/gu/mediaservice/lib/aws/DynamoDBTest.scala]: I'd like to move more of the none IO stuff out to test as well.

File: collections/app/controllers/ImageCollectionsController.scala (hunk 2 above)
Comment (jamesgorrie) [inline collections/app/controllers/ImageCollectionsController.scala]: At the moment we're sending over the whole collection - so no diffing, just set `image.collections = collections`.

Comment (jamesgorrie) [inline collections/app/controllers/ImageCollectionsController.scala]: With deleting we'll do that same, just send the new list [e.g.](https://github.com/guardian/grid/compare/image-collections...image-collection+thrall-notification?expand=1)

Comment (theefer) [inline collections/app/controllers/ImageCollectionsController.scala]: I must have misunderstood what `onlyLatest` does then?

Comment (jamesgorrie) [inline collections/app/controllers/ImageCollectionsController.scala]: As the actual saved list (in Dynamo) could be:
```JSON
{
    "data": [{
        "pathId": "g2/film",
        "path": [
            "g2",
            "film"
        ],
        "actionData": {
            "author": "test",
            "date": "2015-11-30T21:33:21.956+00:00"
        }
    }, {
        "pathId": "g2/film",
        "path": [
            "g2",
            "film"
        ],
        "actionData": {
            "author": "test",
            "date": "2015-12-07T11:23:06.985+00:00"
        }
    }]
}
```

Only latest makes sure that we only use the latest addition.
The reason for this is to keep list storage simple, and just manipulate it after. It also give us an audit trail.

I might write a test to make it more explanitory.

Comment (theefer) [inline collections/app/controllers/ImageCollectionsController.scala]: Oh I see. OK :+1:

Comment (jamesgorrie) [inline collections/app/controllers/ImageCollectionsController.scala]: [Test added](https://github.com/guardian/grid/commit/337b5ed6901d806f1012d74884e7d5c4e2fdaeb2)

Comment (theefer) [inline collections/app/controllers/ImageCollectionsController.scala]: Thanks!

**Comment by theefer** [comment]
Looks legit :+1:

**Pull request merged**